cpp/src/hash/hashing.cu

```diff
@@ -100,61 +100,73 @@ hash_partition_table(table_view const& input,
     // Compute which partition each row belongs to using bitwise partitioner
     auto partitioner = bitwise_partitioner<hash_value_type>{num_partitions};
     thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
-      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
-
+      hash_iterator + num_rows, row_partitions.begin(), partitioner);
   } else {
     // Compute which partition each row belongs to using modulo partitioner
     auto partitioner = modulo_partitioner<hash_value_type>{num_partitions};
     thrust::transform(rmm::exec_policy(stream)->on(stream), hash_iterator,
-      hash_iterator + num_rows, row_partition_numbers.begin(), partitioner);
+      hash_iterator + num_rows, row_partitions.begin(), partitioner);
   }
 
-  // Scatter input rows to output partitions given the partition map
-  auto const partition_map = column_view(data_type(INT32), num_rows,
-    row_partition_numbers.data().get());
-  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
+  // Initialize gather maps and offsets to sequence
+  auto gather_map = rmm::device_vector<size_type>(num_rows);
+  auto d_offsets = rmm::device_vector<size_type>(num_rows);
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
+    gather_map.begin(), gather_map.end());
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
+    d_offsets.begin(), d_offsets.end());
 
-  // Pad with empty tables if we have less than num_partitions
-  if (output.size() < static_cast<size_t>(num_partitions)) {
-    std::generate_n(std::back_inserter(output), num_partitions - output.size(),
-      [&input]() { return experimental::empty_like(input); });
-  }
+  // Sort sequence using partition map as key to generate gather maps
+  thrust::stable_sort_by_key(rmm::exec_policy(stream)->on(stream),
+    row_partitions.begin(), row_partitions.end(), gather_map.begin());
 
-  return output;
+  // Reduce unique partitions to extract gather map offsets from sequence
+  auto end = thrust::unique_by_key(rmm::exec_policy(stream)->on(stream),
+    row_partitions.begin(), row_partitions.end(), d_offsets.begin());
+
+  auto output = experimental::detail::gather(input, gather_map.begin(), gather_map.end(),
+    false, false, false, mr, stream);
+
+  // Copy offsets from device to vector, padded to num_partitions
```

Comment:
This will effectively move any empty partitions to the end of list. I thought this didn't really matter so I didn't bother to do extra computation to insert the empty bins in the middle where they would be otherwise.

Reply:
I think this is going to be a problem because the actual index of the partition matters, not just that they are divided up.  The use case is deciding which host to send a particular row to, based off of the key for that row.  Think about an equi-join where we want all of rows with the same keys to show up on the same host.  I need a guarantee that no matter what else is in the table with a given row it will always hash to the exact same partition number, assuming that the input number of partitions is always the same.

Reply:
Ah, I see your point! Ok, I will add an extra step to insert empty bins in the middle if needed to keep the partition number.

Reply:
Just to be clear, it doesn't matter what order the rows are in within the partition, just that the rows always end up in the same partition based on their hash, correct?

Reply:
Yes that is correct.  The order of the rows in the partition is not important, just the partition number.

Reply:
I tested with the latest on this branch and it looks good to me I am +1